41 integrate local database

mlanima opened the pull request:
Integrated local databse.

Changes:

- integrated Prisma ORM
- created prisma models (database tables)
- created seed for database that loads data from lib/data we used earlier as static data
- added api route to retrieve ducks
- switched from use of static ducks to ducks from database
- added scripts to communicate with prisma from the root

To test, firstly you need to generate prisma client (client will be generated on npm install automatically):

`npm run prisma:generate`

The created client will use database provided in .env of packeges/web folder. Yo can use compose.yaml to set up the database from a container.

Then you should create database tables by migrating them:
`npm run prisma:migrate`

It is possible to provide the database with sample data from lib/data by running:
`npm run prisma:seed`

Prisma ORM Docs - [https://www.prisma.io/docs/](Docs)

mlanima committed:
init prisma

mlanima committed:
fetch ducks in home page insted of static data

mlanima committed:
delete env* from local .gitignore

untbu commented:
Regarding point 2 and 7, you could use the following prepare script in the root `package.json`: `"prepare": "dotenv -e packages/web/.env -- prisma generate"`. This runs automatically during `npm install`.
4. Maybe someone with more experience with Next should comment on this.
5. If prisma does lazy loading, everything is fine here
6. I don't want to remove the dummy data, but currently the dummy data is in the file `prisma/data.ts` and `packages/web/src/data/data.ts`

mlanima committed:
resolve problems

mlanima committed:
move docker compose, add migrate to prepare

mlanima committed:
gitignore generated

mlanima committed:
Merge branch 'main' into 41-integrate-local-database

mlanima committed:
resolve ci/cd problem

mlanima commented:
I fixed the problems we discussed.

- I added docker-compose for database. I dont know if it really needed but I used it for local connection.
- now before running the project you should provide database connection (.env) and on npm install prisma client will be generated. Then you should run `npm run prisma:migrate` to create tables at database. If you want it is possible to create dummy ducks from /data folder `npm run prisma:seed`.

But i have some problems with ci/cd due CLRF/LF and I dont know how to fix it...

Comet1903 commented on packages/web/.env.example:
Add a line break to the end of the file

Comet1903 commented on package.json:
```diff
@@ -7,13 +7,17 @@
+    "prisma:migrate": "npm run prisma:migrate --workspace=@sose2025/web",
+    "prisma:seed": "npm run prisma:seed --workspace=@sose2025/web",
+    "prepare": "npm run prepare --workspace=@sose2025/web"
+
```
Remove this line break

mlanima committed:
resolbe esling problems

Comet1903 commented on packages/web/prisma/migrations/20250612155634_init/migration.sql:
Some types are very open. I would suggest restricting the data types to specific contents.

Here's an example:
In the table `Producers`, the field `email` is of type `TEXT`. Since there are limits regarding to the size of email addresses, a maximum of about 320 characters is enough. Also, it would be nice if the database could check whether there is a valid email address in an email field (maybe use a function that checks that with regex?)

Since I’m not a huge fan of fully normalizing databases, I personally wouldn’t go all the way, but:

The attribute `address` should not be a single attribute. You should split it up into `country`, `region`, etc. Otherwise, we may not be able to select every customer in, for example, Berlin, if you don’t change that.

Personally: All of the `CONSTRAINT "xxx_pkey" PRIMARY KEY ("id")` can also be directly be written right after `"id" SERIAL...` this would look like this:

```
...SNIP...
    "id" SERIAL NOT NULL PRIMARY KEY,
...SNIP...
```

And a short question about all of the indexes: are they really all needed? Since they consume some space, I’d only add them if we currently need them or are sure that we will.

mlanima replied:
I changed the type to Varchar(130) natively. Thank you for this posint. Hovewer, when working with prisma as I understood most constraints are expected to be set by input validation on a server. It should be noted that migrations are automatically generated, so it doesnt really important how and in which order sql instruction are written in to the file.

mlanima committed:
remove use of static data

mlanima committed:
import ignoreGlobal and run prettier

Comet1903 reviewed (commented):
I took a look at the changes and had a few concerns. Feel free to review my comments and let me know what you think. Most of my feedback relates to the database structure and the Docker Compose file.

mlanima committed:
limit email length, rename compose file, add linebreaks where needed

no4any reviewed (changes requested):
Missing documentation in README.md on how to start environment.

mlanima committed:
write README on how to start enviroment and use prisma

mlanima commented:
> Missing documentation in README.md on how to start environment.

Thank you for your comment. Now there is docu in README on how to start the enviroment and use prisma

untbu committed:
edit README.md

Co-authored-by: Julian Horn <mail@julian-horn.net>

Comet1903 commented on packages/web/src/app/api/ducks/route.ts:
```diff
@@ -0,0 +1,42 @@
+// GET /api/ducks
+export async function GET() {
+    try {
+        // Fetch ducks with related data
+        const ducks = await prisma.duck.findMany({
+            include: {
+                brand: true,
+                origin: true,
+                producer: true,
+            },
+        });
+
+        // Map data from Prisma to fit RubberDuckSchema
+        const formatted = ducks.map((duck) => ({
+            id: duck.id,
+            name: duck.name,
+            price: duck.price,
+            color: duck.color, // Prisma enum as string → Zod enum accepts string
+            material: duck.material,
+            size: duck.size, // Prisma enum as string → Zod enum accepts string
+            brand: duck.brand?.name ?? null, // optional()
+            origin: duck.origin.name,
+            producer: duck.producer.name,
+            weight: duck.weight,
+            description: duck.description,
+        }));
```
I think, if you select all ducks with all of their brands, origins and producers, then map them to the `RubberDuckSchema` there will be a lot of over fetching. Is that right?

Is the query you wrote above an equivalent to something like this? 
```sql
SELECT * FROM Duck
INNER JOIN Brand ON Duck.brandId = Brand.id
INNER JOIN Origin ON Duck.originId = Origin.id
INNER JOIN Producer ON Duck.producerId = Producer.id
```

Where actually something like this is needed:

```sql
SELECT Duck.*, Brand.name AS brand, Origin.name AS origin, Producer.name AS producer FROM Duck
INNER JOIN Brand ON Duck.brandId = Brand.id
INNER JOIN Origin ON Duck.originId = Origin.id
INNER JOIN Producer ON Duck.producerId = Producer.id
```

Those two queries I wrote are more likely to be pseudocode. I did't test them.

Comet1903 commented on packages/web/src/app/items/[id]/page.tsx:
These changes doesn't work, since there are some fields that were not selected by the query.

Resolve the missing relations and this should work as expected.

Comet1903 commented on packages/web/src/app/NewestProducts.component.tsx:
```diff
@@ -1,16 +1,33 @@
+    const [products, setProducts] = useState<RubberDuck[]>([]);
+
+    useEffect(() => {
+        fetch("/api/ducks")
+            .then((res) => res.json())
+            .then((data) => {
+                setProducts(DucksArraySchema.parse(data));
+            })
+            .catch((error) => {
+                console.error("Error fetching products:", error);
+                setProducts([]); // Set to empty array on error
+            });
```
Since you chose to do it the next.js way in `packages/web/src/app/items/[id]/page.tsx`, why is this an api call through the javascript fetch api instead of the typical next.js function?

Comet1903 reviewed (changes requested):
I took another look into this pull request, since I was reviewing other pull requests and found some other things that may be changed. While looking again and again in the migrations, I also found other things to address.

Maybe you can take a look into it and fix that or tell my why not. Especially why you did't change the values of type `TEXT` to `VARCHAR(n)` as I suggested in an earlier comment on this pull request.

mlanima committed:
resolve overfetchin at /ducks route

Mimamo03 committed:
Fix Databse Structure

Mimamo03 committed:
Fix For linter

Mimamo03 committed:
Fix database Schema

mlanima committed:
resolve problems when displaying ducks

mlanima committed:
remove unused api routes

mlanima committed:
remove redundancy at ul

mlanima committed:
add admin table to db

mlanima committed:
apply format

mlanima committed:
Merge branch 'main' into 41-integrate-local-database

mlanima committed:
use client for displaying ducks

Mimamo03 committed:
Fix for 404 in getDuck

mlanima committed:
use server actions for fetching

mlanima committed:
Merge branch 'main' into 41-integrate-local-database

mlanima committed:
resolve build problem

mlanima committed:
format

mlanima commented:
Api calls for data fetching were replaced with server actions and main was merged. Everything works good so far.

eliasbnr commented on packages/web/prisma/schema.prisma:
```diff
@@ -0,0 +1,171 @@
+    @@index([color])
+}
+
+model Order {
+    id          Int         @id @default(autoincrement())
+    createdAt   DateTime    @default(now()) @db.Timestamp(0)
+    totalPrice  Int
+    isCompleted Status      @default(pending)
+    items       OrderItem[]
+    customerId  Int         @unique
+
+    customer Customer @relation(fields: [customerId], references: [id])
+
+    paypalTransaction PaypalTransaction?
+
+    // for performance when ordering by creation date or completion status
+    @@index([createdAt])
+    @@index([isCompleted])
+}
+
+model Customer {
+    id               Int    @id @default(autoincrement())
+    firstName        String @db.VarChar(255)
+    lastName         String @db.VarChar(255)
+    email            String @db.VarChar(320)
+    phone            String @db.VarChar(20)
+    homeAddressId    Int // Optional field syntax: `Int?`
+    billingAddressId Int // Required field syntax: `Int`
+
+    homeAddress    Address? @relation("HomeAddressRelation", fields: [homeAddressId], references: [id])
+    billingAddress Address  @relation("BillingAddressRelation", fields: [billingAddressId], references: [id])
+
+    order Order?
+}
```
The Order model should allow multiple orders per customer:

```
model Order {
    id          Int         @id @default(autoincrement())
    createdAt   DateTime    @default(now()) @db.Timestamp(0)
    totalPrice  Int
    isCompleted Status      @default(pending)
    items       OrderItem[]
    customerId  Int         // Remove @unique here

    customer Customer @relation(fields: [customerId], references: [id])
    // ... rest of the model
}

model Customer {
    // ... other fields
    orders Order[] // Change from order Order? to orders Order[]
}
```

eliasbnr commented on packages/web/src/lib/paypal.ts:
```diff
@@ -80,7 +80,9 @@ export class PayPalApiService {
         const items: Item[] = [];
         let total = 0;
         for (const entry of cart) {
-            const item = rubberDuckData.find((x) => x.id == entry.id);
+            const item = await prisma.duck.findUnique({
+                where: { id: entry.id },
+            });
```
This creates N+1 query problem (one database query per cart item).
Recommendation: Use `findMany` with `where: { id: { in: cartIds } }`

```
// Instead of N+1 queries:
for (const entry of cart) {
    const item = await prisma.duck.findUnique({ where: { id: entry.id } });
}

// Use this approach:
const cartIds = cart.map(entry => entry.id);
const items = await prisma.duck.findMany({
    where: { id: { in: cartIds } }
});
const itemsMap = new Map(items.map(item => [item.id, item]));
```

eliasbnr commented on packages/web/src/app/NewestProducts.component.tsx:
```diff
@@ -70,20 +69,28 @@ const NoResults = () => (
+    const [products, setProducts] = useState<RubberDuck[]>([]);
+    const [filteredProducts, setFilteredProducts] = useState<RubberDuck[]>([]);
+
+    // Fetch products on mount
+    useEffect(() => {
+        getDucks()
+            .then((ducks) => setProducts(ducks))
+            .catch(notFound);
+    }, []);
```
notFound() in useEffect might not work as expected. Add proper error state handling in the component.

```
// In components, use proper error states:
const [error, setError] = useState<string | null>(null);

useEffect(() => {
    getDucks()
        .then(setProducts)
        .catch((err) => {
            console.error('Failed to load products:', err);
            setError('Failed to load products');
        });
}, []);
```

eliasbnr commented on packages/web/src/app/api/ducks/[id]/route.ts:
```diff
@@ -0,0 +1,10 @@
+    const duck = await prisma.duck.findUnique({ where: { id: Number(id) } });
+
+    return NextResponse.json(duck);
+}
```
Add Input Validation to API Routes:

```
import { z } from 'zod';

const ParamsSchema = z.object({
    id: z.string().transform(val => {
        const num = parseInt(val, 10);
        if (isNaN(num)) throw new Error('Invalid ID');
        return num;
    })
});

export async function GET(req: NextRequest, { params }: { params: Promise<{ id: string }> }) {
    try {
        const { id } = ParamsSchema.parse(await params);
        const duck = await prisma.duck.findUnique({ where: { id } });
        
        if (!duck) {
            return NextResponse.json({ error: 'Duck not found' }, { status: 404 });
        }
        
        return NextResponse.json(duck);
    } catch (error) {
        return NextResponse.json({ error: 'Invalid request' }, { status: 400 });
    }
}
```

mlanima committed:
remove unused api

takh86 reviewed (commented):
super
